ARROW-5610: [Python] define extension types in Python

jorisvandenbossche opened the pull request:
https://issues.apache.org/jira/browse/ARROW-5610

jorisvandenbossche commented on cpp/src/arrow/python/extension_type.cc:
```diff
@@ -193,5 +193,105 @@ Status UnregisterPyExtensionType() { return UnregisterExtensionType(kExtensionNa
+// Define a generic extension type class that takes the name as constructor
+// parameter
+
+GenericExtensionType::GenericExtensionType(std::shared_ptr<DataType> storage_type,
```
This (and below) is mostly the same as the code above for `PyExtensionType` with only a few changes (if we decide to keep both, I should look into deduplicating the code)

BryanCutler replied:
Is it necessary to have the `GenericExtensionType` also? My initial thoughts were that `PyExtensionType` could be modified to use other names and use the registry

jorisvandenbossche replied:
There are some differences that I initially thought might be difficult to combine in a single class. But looking at it now again, I see the main difference is (apart from some checks of the constant name):

- `Serialize` always calls the instance (as it could be parametrized, so the serialization could be different for each instance), while in the current PyExtensionType this is a constant (a pickle of the class, which does not depend on the instance, only on the subclass). I suppose always calling it is not problematic here, also in the case where it would not be needed.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -217,3 +217,50 @@ def test_ipc_unknown_type():
     assert arr.type == ParamExtType(3)
+
+
+class PeriodType(pa.GenericExtensionType):
```
this is an example of how the API would look like for people implementing an extension type in Python.

Here, as initial example, I took the "period" type from pandas (which is currently not supported by pyarrow out of the box), which is basically integer values with a frequency.

dhirschfeld replied:
Unfortunately I'm having trouble understanding the intended interface from this example as it's too simple - AFAICS it effectively boils down to:
```python
    def __arrow_ext_serialize__(self):
        return self.freq.encode()

    @classmethod
    def __arrow_ext_deserialize__(cls, storage_type, serialized):
        freq = serialized.decode()
        return PeriodType(freq)
```
i.e. it's really just passing a python string and then using the same string to instantiate the class - no serialization or other involvement with arrow that I can see.

I assume the api of the serialize method is is:
```python
def __arrow_ext_serialize__(self) -> bytes:
    ...
```
(documentation using type hints to show the api would be extremely helpful - especially since the actually implementation will be hidden in a compiled module)

I can decompose my python types into (possibly nested) dictionaries mapping names to simple types where simple types include stings, numbers, timestamps or arrays/dsataframes of those types. How do I then turn that into a single bytearray to return from `__arrow_ext_serialize__`?

jorisvandenbossche replied:
> i.e. it's really just passing a python string and then using the same string to instantiate the class - no serialization or other involvement with arrow that I can see.

Yes, that is correct. But what aspect do you not understand about this example? (It's simple, but still an actual real world use case (that I intend to implement in pandas), so that it can be made simple seems like a plus)

> I can decompose my python types into (possibly nested) dictionaries mapping names to simple types where simple types include stings, numbers, timestamps or arrays/dsataframes of those types. How do I then turn that into a single bytearray to return from __arrow_ext_serialize__?

Can you give a more concrete example of what you are trying to achieve here?

dhirschfeld replied:
I understand the example fine - I just don't know how to use it to implement anything not trivial.

The example is trivial because it's not using arrow at all. It serializes a python string by calling the `encode` method and `deserializes` by calling `decode`. This works for strings but doesn't say anything about how to convert any other python object (which doesn't have handy `encode` /`decode` method)  into bytes.

jorisvandenbossche commented on python/pyarrow/public-api.pxi:
```diff
@@ -96,6 +98,9 @@ cdef api object pyarrow_wrap_data_type(
             return cpy_ext_type.GetInstance()
+        # elif ext_type.extension_name() == period_ext_name:
+        #     cgen_ext_type = <const CGenericExtensionType*> ext_type
+        #     return cgen_ext_type.GetInstance()
```
For some reason, this doesn't work yet to get an actual instance back (it gives `None`). 
Also, I would need a way to check that the C type is actually a GenericExtensionType in the `elif` check.

BryanCutler commented on cpp/src/arrow/python/extension_type.h:
```diff
@@ -73,5 +73,61 @@ ARROW_PYTHON_EXPORT Status RegisterPyExtensionType(const std::shared_ptr<DataTyp
+    if (other_ext.storage_type() != this->storage_type()) {
+      return false;
+    }
+    return true;
```
could this call a Python method to compare type parameters? That way the user could override the Python method and check parameters, so that they could register the extension once but have multiple parametrized types. The default would to just return True.

BryanCutler commented on python/pyarrow/types.pxi:
```diff
@@ -614,6 +614,76 @@ cdef class UnknownExtensionType(ExtensionType):
+        check_status(self.cgen_ext_type.SetInstance(self))
+
+    def __arrow_ext_serialize__(self):
+        # default for non-paramtrized type is to have no metadata
```
can it default to use pickle format?

jorisvandenbossche replied:
> can it default to use pickle format?

In order to combine it with the other class? (which already uses pickle)

I personally don't really like pickle as default (as it makes it python specific), but if we want to combine both approaches in a single class, then yes that might be the default

BryanCutler commented on python/pyarrow/types.pxi:
```diff
@@ -614,6 +614,76 @@ cdef class UnknownExtensionType(ExtensionType):
+        # Store weakref on C++ type instance
+        check_status(self.cgen_ext_type.SetInstance(self))
+
+    def __arrow_ext_serialize__(self):
```
is there a reason why not to use `def serialize()` and `def deserialize()`? It would be a little nicer for the end user to override

wesm replied:
I commented above about this. I had thought that serialize/deserialize would be intuitive -- now that these are pseudo-public APIs we might take the opportunity to change. Interested in @pitrou's thoughts

BryanCutler commented on python/pyarrow/types.pxi:
```diff
@@ -614,6 +614,76 @@ cdef class UnknownExtensionType(ExtensionType):
+def unregister_extension_type(type_name):
+    cdef:
+        c_string c_type_name = tobytes(type_name)
+    check_status(UnregisterGenericExtensionType(c_type_name))
```
do you need to unregister on the python side?

jorisvandenbossche replied:
Strictly speaking it is not needed for the reason I started the registry (to ensure we unregister it on the C++ side), as we do that in a try except, but it's probably a good idea to do that anyway, yes. That seems cleaner.

jorisvandenbossche commented:
@BryanCutler I updated the pull request, and now removed the C++ GenericExtensionType (moved the needed functionality into the existing C++ PyExtesionType). 

On the python side, for now I kept the two concrete base classes `ExtensionType` (based on pickle to (de)serialize, with fixed name, no need to manually register) and `GenericExtensionType` (with variable name and custom serialize, need to register). They give some different functionality to the user subclassing them, and not fully sure if or how to unify them. 
(but if we keep them separate, a better name for "GenericExtensionType" would be welcome)

jorisvandenbossche commented:
cc @pitrou if you some time, a review would always be welcome since you wrote the original version

wesm commented on python/pyarrow/public-api.pxi:
```diff
@@ -93,8 +98,8 @@ cdef api object pyarrow_wrap_data_type(
-        if ext_type.extension_name() == PyExtensionName():
-            cpy_ext_type = <const CPyExtensionType*> ext_type
+        cpy_ext_type = dynamic_cast[_CPyExtensionTypePtr](ext_type)
+        if cpy_ext_type != nullptr:
```
I didn't realize you could use `nullptr` in Cython is that new?

jorisvandenbossche replied:
> I didn't realize you could use nullptr in Cython is that new?

It's imported in lib.pxd with `from libcpp cimport nullptr`, and used in several places (but no idea if this is a new cython feature or not)

jorisvandenbossche commented:
> Do we want to try calling these PyExtensionType and NativeExtensionType or something?

If we keep the two distinct Python classes, then that seem good names. 

But, thinking a bit more about it, it is actually *possible* to only have a single python class (that implementors need to subclass). The default "extension_name" could be "arrow.py_extension_type" (but the type implementor can override them), and the default serialize/deserialize can be based on pickle (but the type implementor can override them)

But although possible, I am not sure it necessarily leads to a clearer API (but good examples on the different ways how you can subclass it can do a lot of course).

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+For example, we could define a custom UUID type for 128-bit numbers which can
+be represented as ``FixedSizeBinary`` type with 16 bytes.
+Using the first approach, we create a `UuidType` subclass, and implement the
+``__reduce__`` dunder method to ensure the class can be properly pickled::
```
Or simply "method"...

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+        def __init__(self):
+            pa.GenericExtensionType.__init__(self, pa.binary(16), "my_package.uuid")
+
+        def __arrow_ext_serialize__(self):
```
I think it's fine to keep them as-is. It hints that they're not meant to be called directly by the user; rather, Arrow is calling them.

jorisvandenbossche replied:
I agree with @pitrou here. For me, it seems similar to methods like `__eq__` etc that you can define on your class, as those are also not meant to be called directly by the end-user, but are called by low-level code.

jorisvandenbossche replied:
The name might be misleading, but as far as I understand, this is not meant to be a proper standalone serialization method. 
With the current metadata specification, the serialized bytes don't need to completely describe the type (and thus are not necessarily fully descriptive to result in the type when deserialized), but only in combination with a type name and a registry of types. 

The result of the "serialize" method is stored in the "ARROW:extension:metadata" key of the field's metadata. So it is more "serialized form of the type's metadata", than actually a "serialized form of the type".

Of course, it is not really specified (as far as I know), how this metadata field needs to be used, and it could also hold a serialization of the full type. Like the pickle-based PyExtensionType does, but that is specific to that specific implementation.

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+    >>> uuid_type.extension_name
+    'my_package.uuid'
+
+    >>> pa.lib.register_extension_type(uuid_type)
```
Hmm, does the user really need to call the registration function themselves?
If so, perhaps allow using it as a class decorator?
```python
@pa.register_extension_type
class UUIDType(...): #etc.
```

dhirschfeld replied:
Could sub-classing not automatically register the type? Is there a reason you'd want to subclass `GenericExtensionType` but *not* register it?

jorisvandenbossche replied:
> If so, perhaps allow using it as a class decorator?

Yes, that could be provided (assuming that the constructor (`__init__`) has defaults set so it can be called to create an instance, as the current registration needs an instance.

> Could sub-classing not automatically register the type?

This might be possible with a metaclass. I can take a look if I can get it to work in cython.

dhirschfeld replied:
If it adds some overhead though I'd actually prefer to manually register the class. I've got some lightweight dataclasses I'd like to be arrow serializable without introducing too much overhead.

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+the extension type as a "uuid" type, if it has implemented its own extension
+type to receive it.
+
+TODO: add example of parametrized type
```
That sounds important indeed.

pitrou commented on cpp/src/arrow/python/extension_type.cc:
```diff
@@ -115,14 +119,21 @@ bool PyExtensionType::ExtensionEquals(const ExtensionType& other) const {
 }
 
 std::string PyExtensionType::Serialize() const {
+  PyAcquireGIL lock;
```
Is there a particular reason you changed this implementation?

jorisvandenbossche replied:
> Is there a particular reason you changed this implementation?

I wanted to answer:

> Yes, because now the "stored" `serialized_` might not be up to date (eg in case of parametrized extension type, it is registered with a certain parametrization, but the actual instance can have a different parametrization). So you actually need to call the serialize method of the instance, instead of returning the stored `serialized_`. 

But now I think about it, I don't know anymore why I added this .. Because the `serialized_` is still set in `SetInstance`, which is called in the `__init__` of the python ExtensionType. So for a given instance, the stored `serialized_` is always correct. 

Unless your python ExtensionType is mutable (eg with the PeriodType example, that you set the `freq` parameter to a different value after creating the type instance). But I don't think we want to support this?  
(although there is not directly a way to prevent users doing this?)

pitrou replied:
No, types are meant to be immutable.

jorisvandenbossche replied:
> No, types are meant to be immutable.

I should probably mention that in the docs then. 
And actually update the example. The two parametrized examples we have in the tests (and which I was planning to add to the docs as well) simply set a class attribute in the `__init__`, which means this attribute is mutable. So for a good example, I should probably turn them into a property (and store the actual attribute in a private attribute)

pitrou commented on cpp/src/arrow/python/extension_type.cc:
```diff
@@ -115,14 +119,21 @@ bool PyExtensionType::ExtensionEquals(const ExtensionType& other) const {
+  OwnedRef inst(GetInstance());
+
+  std::string serialized;
+  SerializeExtInstance(inst.obj(), &serialized);
```
Also, should not ignore the return value of `SerializeExtInstance`...

pitrou commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -217,3 +217,130 @@ def test_ipc_unknown_type():
+    # setup
+    period_type = PeriodType('D')
+    pa.lib.register_extension_type(period_type)
+    yield
```
`try..finally` here? Or is that not necessary with pytest?

jorisvandenbossche replied:
> try..finally here? Or is that not necessary with pytest?

AFAIU, this is not needed with pytest (it's like a unittest teardown method, the part after yield should be called in the end regardless of whether the test raised an error or not)

pitrou commented:
I posted some comments. I don't think there is a point in combining the two concrete ExtensionType classes.

dhirschfeld commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+This can now be used to create arrays and tables holding the extension type::
+
+    >>> uuid_type = UuidType()
+    >>> uuid_type.extension_name
```
Should `extension_name` be a private attrribute so the user doesn't *have* to have it as part of their classes' public api?
i.e. `s/extension_name/_extension_name` and if the user wants to opt in they can implement a property:
```python
@property
def extension_name(self):
    return self._extension_name
```
Similarly for `storage_type`.

Alternatively perhaps they can be "namespaced" - e.g.
```python
uuid_type.arrow.extension_name
uuid_type.arrow.storage_type
```

jorisvandenbossche replied:
@dhirschfeld is there a reason you wouldn't want it publicly? 
This is currently matching the C++ API where the extension name and storage type are methods on the type class. Those two properties are also quite fundamental to the concept of "extension types" in arrow, so I would personally not hide them.

dhirschfeld replied:
It prevents me having my own `storage_type` attribute which means something different in the context of my library.

...I don't actually have a `storage_type` attribute so it's just a hypothetical argument and thinking about it some more you could namespace it yourself:
```python
    @property
    def arrow_storage_type(self):
        return super().storage_type
```

dhirschfeld replied:
The principle is that I don't necessarily want my users to know the classes they're instantiating are being serialized by arrow - that's just an implementation detail similar to how the pickle protocol is defined using a private `__reduce__` method rather than a public `pickle` method

i.e. implementing the pickle protocol doesn't pollute the public api of your class with random attributes or methods.

> *Those two properties are also quite fundamental to the concept of "extension types" in arrow*

They're fundamental to arrow but irrelevant to my users. My users just care that they can save their instances so they can be recreated later and for me, arrow will hopefully allow the recreation to happen in Python, R or C# - i.e. the benefit over just using pickle is potentially enable cross-language interop.

dhirschfeld commented:
> *Can you give a more concrete example of what you are trying to achieve here?*

@jorisvandenbossche - I've got a demo serialization framework set up using the existing capability and I'd like to update it to this new methodology. I'm sure it's possible but I can't figure out how from this PR.

I've knocked up a demo notebook over at:
https://anaconda.org/dhirschfeld/arrowserialization/notebook

Over there I implement an arrow-serializable class `MyClass`:

```python
class MyClass(ArrowSerializable):
    def __init__(self, scalar, array, timestamp, dataframe):
        self.scalar = scalar
        self.array = array
        self.timestamp = timestamp
        self.dataframe = dataframe

    def __getstate__(self):
        return dict(
            scalar=self.scalar,
            array=self.array,
            timestamp=self.timestamp,
            dataframe=self.dataframe,
        )
```
![image](https://user-images.githubusercontent.com/881019/63734375-5a408f80-c8bf-11e9-9dda-dfaee0e9d671.png)

What I'd like to know is, is such a class intended to be serializable by this new framework and if so, how?!

What I'm hoping for is that by subclassing from `pa.GenericExtensionType` the performance will be better:

![image](https://user-images.githubusercontent.com/881019/63734612-4cd7d500-c8c0-11e9-9070-3c7fc873914e.png)

...and that I will be theoretically able to read python serialized objects from either R or C#! :rocket:

jorisvandenbossche commented:
@dhirschfeld thanks for the elaborate feedback!

But, I think there is some confusion around the purpose of the extension types. 
The example you show involves the serialization functionality of pyarrow. This is a somewhat distinct part of pyarrow than the Table/RecordBatch/Array functionality that is part of the standardized memory layout and metadata format. (I also think the serialization functionality of pyarrow is currently restricted to pyarrow, and not interoperable with other Arrow implementations)

So the ExtensionTypes come into play if you want to define a custom interpretation of a, *Arrow array*, not of any custom pyarrow-serializable object. If your custom objects can be converted to a Arrow array, then you could write an ExtensionType to hold some metadata information about this custom object. But from the example you show, it does not seem that you try to convert this custom instance to an Array?

jorisvandenbossche commented:
There was recently an issue opened about the need to clarify the relationship between the serialization framework and the IPC messages: https://issues.apache.org/jira/browse/ARROW-6336

dhirschfeld commented:
> If your custom objects can be converted to a Arrow array, then you could write an ExtensionType to hold some metadata information about this custom object. But from the example you show, it does not seem that you try to convert this custom instance to an Array?

I convert my classes to Python dictionaries of simple types... so I guess this isn't a use-case `pa.GenericExtensionType` is designed to support?

i.e. this framework isn't designed to serialize arbitrary classes such as:
```python
In [58]: @dataclass
    ...: class DatetimeInterval(object):
    ...:     start: date
    ...:     end: date
    ...:     closed: str = 'left'
```
?
> *Does that make sense?*

I guess I'm still confused - in what way the `PeriodType` an Arrow array?

> if you have some pointers on how to clarify this in the docs

All I could suggest is another possibly slightly more complicated example :/

Thanks for the link to the JIRA - I'm watching that now as I'm also unsure of the relationship between IPC and serialization.

jorisvandenbossche commented:
> in what way the PeriodType an Arrow array?

It's not an Arrow array in itself, but it is defining a *data type*  that an Arrow array can hold. In other words, you can create an array that has a `PeriodType` type.

In this example, I am defining a period extension type / array (which is basically an integer array with some metadata, i.e. the frequency (how to interpret the integers)). Potentially, this can be the target to convert a pandas period dtype column into a Arrow array, when converting a pandas DataFrame holding such a period dtype column to an Arrow table.

jorisvandenbossche commented:
> i.e. this framework isn't designed to serialize arbitrary classes such as:

It is not directly, but, it could be used if you would have an array of such DatetimeInterval objects, and you want to represent this array of custom objects with one of the native types in Arrow (eg for the example, this could be a StructArray).

jorisvandenbossche commented on cpp/src/arrow/python/extension_type.cc:
```diff
@@ -182,14 +193,25 @@ Status PyExtensionType::FromClass(std::shared_ptr<DataType> storage_type, PyObje
+                                         PyObject* typ,
+                                         std::shared_ptr<ExtensionType>* out) {
+  Py_INCREF(typ);
+  out->reset(new PyExtensionType(storage_type, extension_name, typ));
```
Done

jorisvandenbossche commented on cpp/src/arrow/python/extension_type.h:
```diff
@@ -46,6 +46,9 @@ class ARROW_PYTHON_EXPORT PyExtensionType : public ExtensionType {
   // For use from Cython
   static Status FromClass(std::shared_ptr<DataType> storage_type, PyObject* typ,
                           std::shared_ptr<ExtensionType>* out);
+  static Status FromClassAndName(std::shared_ptr<DataType> storage_type,
```
I ended up combining `FromClassAndName` into `FromClass`, as the extension name can be set from the python side.

jorisvandenbossche commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+
+* Subclassing :class:`ExtensionType`: the (de)serialization is based on pickle.
+  This is a good option for an extension type that is only used from Python.
+* Subclassing :class:`GenericExtensionType`: this allows to give a custom
```
I ended up renaming it to `PyExtensionType` and just `ExtensionType`, as to avoid code duplication, I actually made the pickle-based `PyExtensionType` a subclass of the native ExtensionType

jorisvandenbossche commented on docs/source/python/extending_types.rst:
```diff
@@ -0,0 +1,115 @@
+      718747F48E5F4058A7261E2B6B228BE8,
+      7FE201227D624D96A5CD8639DEF2A68B,
+      C6CA8C7F95744BFD9462A40B3F57A86C
+    ]
```
> Do you want to have this example be generated at Sphinx build time using .. code-block: ipython?

Ideally yes, I suppose (although I need to check if the class definition works with ipython). But for reviewing now I thought it would be easier to see the actual content.

jorisvandenbossche commented:
I further updated the docs to include a section on parametrized extension types

@pitrou @wesm @BryanCutler more comments / feedback on the new version?

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -44,3 +44,162 @@ define the ``__arrow_array__`` method to return an Arrow array::
+The above example used a fixed storage type with no further metadata. But
+more flexible, parametrized extension types are also possible.
+
+The example given here implements an extension type for the pandas "period"
```
Add a link to Pandas or its period datatype? Also, should "Pandas" be capitalized?

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -44,3 +44,162 @@ define the ``__arrow_array__`` method to return an Arrow array::
+The example given here implements an extension type for the pandas "period"
+data type, represententing time spans (e.g., a day, a month, a quarter, etc).
+It is stored as an int64 array which is interpreted as the number of time spans
+of the given frequency since 1970.
```
"period" rather than "frequency"?

pitrou commented on docs/source/python/extending_types.rst:
```diff
@@ -44,3 +44,162 @@ define the ``__arrow_array__`` method to return an Arrow array::
 
 The ``__arrow_array__`` method takes an optional `type` keyword which is passed
 through from :func:`pyarrow.array`.
+
```
I realize this document doesn't talk about registration. Also show a simple IPC example?

pitrou commented on python/pyarrow/types.pxi:
```diff
@@ -608,12 +663,54 @@ cdef class UnknownExtensionType(ExtensionType):
         return self.serialized
 
 
+_PYTHON_EXTENSION_TYPES_REGISTRY = []
```
Nit, but the UPPERCASE looks unnecessary.

pitrou commented on python/pyarrow/types.pxi:
```diff
@@ -608,12 +663,54 @@ cdef class UnknownExtensionType(ExtensionType):
+        raise TypeError("Only extension types can be registered")
+
+    # register on the python side
+    _PYTHON_EXTENSION_TYPES_REGISTRY.append(_type)
```
Should be done after the C++ registration?

pitrou commented:
Will merge. I think we can refine later if desired.
